src/node_ffi.cc

```diff
@@ -0,0 +1,244 @@
+#include "node_ffi.h"
+#include "env-inl.h"
+#include "ffi.h"
+#include "node_binding.h"
+#include "v8-array-buffer.h"
+#include "v8-function-callback.h"
+#include "v8-primitive.h"
+#include "util-inl.h"
+#include "node_external_reference.h"
+
+using v8::Context;
+using v8::ArrayBuffer;
+using v8::BigInt;
+using v8::Number;
+using v8::Array;
+using v8::Isolate;
+using v8::String;
+
+namespace node {
+namespace ffi {
+
+static bool lossless = false;
+
+#define ReturnBigInt(isolate, ptr) args.GetReturnValue()\
+  .Set(BigInt::NewFromUnsigned(isolate, (uint64_t)ptr))
+
+void GetLibrary(const v8::FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  bool isNull = true;
+  std::string fname = "";  // TODO(bengl) not this hacky thing
+  if (!args[0]->IsNull()) {
+    CHECK(args[0]->IsString());
+    node::Utf8Value filename(env->isolate(), args[0]);
+    fname = filename.ToString();
+    isNull = false;
+  }
+
+  std::map<std::string, binding::DLib *> libraries =
+    Realm::GetBindingData<FfiBindingData>(args)->libraries;
+  binding::DLib * lib = nullptr;
+  if (libraries[fname] != nullptr) {
+    lib = libraries[fname];
+  } else {
+    if (isNull) {
+      lib = new binding::DLib(binding::DLib::kDefaultFlags);
+    } else {
+      lib = new binding::DLib(fname.c_str(), binding::DLib::kDefaultFlags);
+    }
+    libraries[fname] = lib;
+    bool success = lib->Open();
+    if (!success) {
+      // TODO(bengl) what now?
+    }
+  }
+  ReturnBigInt(env->isolate(), lib);
+}
+
+void GetSymbol(const v8::FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK(args[0]->IsBigInt());
+  binding::DLib * lib = (binding::DLib *)args[0].As<BigInt>()
+    ->Uint64Value(&lossless);
+  CHECK(args[1]->IsString());
+  node::Utf8Value symName(env->isolate(), args[1]);
+  void* symbol = lib->GetSymbolAddress(symName.ToString().c_str());
+  ReturnBigInt(env->isolate(), symbol);
+}
+
+void GetBufferPointer(const v8::FunctionCallbackInfo<Value>& args) {
+  CHECK(args[0]->IsArrayBuffer());
+  void* data = args[0].As<ArrayBuffer>()->Data();
+
+  ReturnBigInt(args.GetIsolate(), data);
+}
```

Comment:
Can this not happen purely on the JS side? Seems to me like it would be faster to avoid the native barrier hop here if possible.

Reply:
None of the Buffer types have pointer addresses available on the JS side.

We could add them to `Buffer`, and special case those? Also, there's definitely an opportunity to cache the address per ArrayBuffer, so that calls to other ArrayBuffer views on the same underlying ArrayBuffer can be handled in JS.